diesel/src/sqlite/connection/app_defined_fn.rs

```diff
@@ -0,0 +1,69 @@
+//! Plumbing for the implementation of the application defined function API
+
+use std::os::raw as libc;
+
+use super::ffi;
+use super::into_sqlite_result::IntoSqliteResult;
+use super::from_sqlite_value::FromSqliteValue;
+
+/// Context is a wrapper for the SQLite function evaluation context.
+#[derive(Debug)]
+pub struct Context<'a> {
+    ctx: *mut ffi::sqlite3_context,
+    args: &'a [*mut ffi::sqlite3_value],
+}
+
+// Context is translated from rusqlite
+impl<'a> Context<'a> {
+    /// Returns the number of arguments to the function.
+    pub fn len(&self) -> usize {
+        self.args.len()
+    }
+
+    /// Returns `true` when there is no argument.
+    pub fn is_empty(&self) -> bool {
+        self.args.is_empty()
+    }
+
+    /// Returns the `idx`th argument as a `T`.
+    ///
+    /// # Failure
+    ///
+    /// Will panic if `idx` is greater than or equal to `self.len()`.
+    pub fn get<T>(&self, idx: usize) -> T
+    where
+        T: FromSqliteValue
+    {
+        let arg = self.args[idx];
+
+        T::from_sqlite_value(arg)
+    }
+}
+
+pub unsafe extern "C" fn free_boxed_value<T>(p: *mut ::std::os::raw::c_void) {
+    let _: Box<T> = Box::from_raw(::std::mem::transmute(p));
```

Comment:
Can Box::from_raw panic, e.g. when p == NULL? Do we need to deal with this?

Reply:
I'd prefer we have this function take `*mut T` and cast the function pointer rather than the other way around. The call to `transmute` is too likely to accidentally cause problems in the future.